### PR TITLE
MODORDERS-100 Create piece on the basis of total quantity

### PR DESCRIPTION
**Pull request by javalkar-varun**
https://issues.folio.org/browse/MODORDERS-100

## Purpose

- Create piece records, even if not creating item records

## Approach

- Calculate total quantity using `compPOL.getCost().getQuantityPhysical()` and `compPOL.getCost().getQuantityElectronic()` as total items and then subtract inventory quantity
- Update tests

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/rest/impl/PutOrderLineByIdHelper.java [hunk 1]
```diff
@@ -199,19 +200,27 @@ public boolean isUpdateSuccessful() {
...
         // Extract item Id's which already associated with piece records
         List<String> existingItemIds = pieces.getPieces()
             .stream()
             .map(Piece::getItemId)
             .filter(StringUtils::isNotEmpty)
             .collect(Collectors.toList());
-
+        
+        // Create pieces for successfully created item Ids
+        if(!itemIds.isEmpty()) {
```
Comment (piotr-kalashuk) [inline src/main/java/org/folio/rest/impl/PutOrderLineByIdHelper.java]: This check does not cover P/E Mix order format when some electronic resources are not expected to be registered in the inventory. The items will be created for physical resources only and the logic will not create piece records for electronic resources.
I propose to enhance the logic and calculate sum of `compPOL.getCost().getQuantityPhysical()` and `compPOL.getCost().getQuantityElectronic()` as total items and then subtract inventory quantity `HelperUtils#calculateInventoryItemsQuantity(pol)`
Ideally the logic should prepare POL content setting `compPOL.getCost().setQuantity(..)` in the beginning of the POL processing so this is always up to date...

Comment (javalkar-varun) [reply src/main/java/org/folio/rest/impl/PutOrderLineByIdHelper.java]: Thanks for your insight @piotr-kalashuk. Just curious, why should we `getQuantityPhysical()` and `getQuantityElectronic()` from `cost`?

Comment (piotr-kalashuk) [reply src/main/java/org/folio/rest/impl/PutOrderLineByIdHelper.java]: Please have a look at https://issues.folio.org/browse/UIOR-121. The Physical and/or Electronic quantities are specified in Cost details and locations' should be based on these figures.
I just checked and see that there is no `compPOL.getCost().setQuantity(..)` according to schema. So the sum should be used all the time to get overall quantity

File: src/main/java/org/folio/rest/impl/PutOrderLineByIdHelper.java [hunk 2]
```diff
@@ -218,6 +219,11 @@ public boolean isUpdateSuccessful() {
                .filter(id -> !existingItemIds.contains(id))
                .forEach(itemId -> futuresList.add(createPiece(poLineId, itemId)));
 
+        // Create pieces when item record does not exists
```
Comment (piotr-kalashuk) [inline src/main/java/org/folio/rest/impl/PutOrderLineByIdHelper.java]: The `createPieces` method will be invoked only if any inventory updates required. Please add additional conditions in the begining of the `updateInventory` method to cover this case.

Comment (javalkar-varun) [reply src/main/java/org/folio/rest/impl/PutOrderLineByIdHelper.java]: Currently, only `updateInventory` is invoking `createPieces` so why do we need a conditional check? No one else is calling `createPieces`?

Comment (piotr-kalashuk) [reply src/main/java/org/folio/rest/impl/PutOrderLineByIdHelper.java]: I mean this check:
```java
    int expectedItemsQuantity = calculateInventoryItemsQuantity(compPOL);
    if (expectedItemsQuantity == 0) {
      logger.debug("PO Line with '{}' id does not require inventory updates", compPOL.getId());
      return completedFuture(null);
    }
```

If nothing required to be created in inventory, the `createPieces`will not be called at all.

**Comment by javalkar-varun** [comment]
@piotr-kalashuk - This branch was far behind the master and I was behaving funky so I have created a new PR https://github.com/folio-org/mod-orders/pull/103. Thank you for your feedback.

**Pull request closed without merging**